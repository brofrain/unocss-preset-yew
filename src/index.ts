// eslint-disable-next-line unicorn/prefer-node-protocol
import fs from 'fs/promises'

import fg from 'fast-glob'
import MagicString from 'magic-string'
import { transformerVariantGroup, type Preset, extractorSplit } from 'unocss'

interface Options {
  /**
   * Glob for Rust files, so utilities can be extracted
   */
  include: string | string[]
}

const CLASS_UTIL_REG =
  /(?<=((classes|uno)!\(\s*(".*"\s*,\s*(\/\/[^\n]*)?\s*)*))"(.*?)"/g

const retrieveClasses = async ({ include }: Options): Promise<Set<string>> => {
  const transformer = transformerVariantGroup()
  const transform = (magicStr: MagicString) =>
    // @ts-expect-error the other params are not used anyway
    transformer.transform(magicStr)

  const filenames = Array.isArray(include)
    ? (await Promise.all(include.map((glob) => fg(glob)))).flat()
    : await fg(include)

  const classes = new Set<string>()

  for (const filename of filenames) {
    const data = await fs.readFile(filename, { encoding: 'utf8' })

    let potentialClasses: string | undefined = data
      .match(CLASS_UTIL_REG)
      ?.join(' ')

    if (!potentialClasses) {
      continue
    }

    const magicPotentialClasses = new MagicString(potentialClasses)
    await transform(magicPotentialClasses)
    if (magicPotentialClasses.hasChanged()) {
      potentialClasses = magicPotentialClasses.toString()
    }

    const extractedClasses = extractorSplit.extract({
      code: potentialClasses,
      original: potentialClasses,
    })

    if (!extractedClasses || 'then' in extractedClasses) {
      continue
    }

    for (const item of extractedClasses) {
      classes.add(item)
    }
  }

  return classes
}

const DEFAULT_OPTIONS: Options = { include: 'src/**/*.rs' }

export const presetYew = (options: Options = DEFAULT_OPTIONS): Preset<any> => ({
  name: 'yew',
  preflights: [
    {
      layer: 'default',
      getCSS: async ({ generator }) => {
        const classes = await retrieveClasses(options)
        const { css } = await generator.generate(classes, {
          preflights: false,
        })
        return css
      },
    },
  ],
})
