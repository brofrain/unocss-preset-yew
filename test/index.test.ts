import { createGenerator, presetWind, type UnoGenerator } from 'unocss'

import { presetYew } from 'src'

describe('index', () => {
  const fakeGenerator = {
    generate: (classes: Set<string>) => ({ css: classes }),
  }

  it('should retrieve correct classes without variant group feature', async () => {
    const preset = presetYew({
      include: 'test/samples/without_variant_groups.rs',
    })

    expect(
      // @ts-expect-error I know what I'm doing
      await preset.preflights[0].getCSS({ generator: fakeGenerator })
    ).toMatchSnapshot()
  })

  it('should retrieve correct classes with variant group feature', async () => {
    const preset = presetYew({
      include: 'test/samples/with_variant_groups.rs',
    })

    expect(
      // @ts-expect-error I know what I'm doing
      await preset.preflights[0].getCSS({ generator: fakeGenerator })
    ).toMatchSnapshot()
  })

  it('should generate utils without variant group feature', async () => {
    const uno: UnoGenerator = createGenerator({
      presets: [
        presetWind(),
        presetYew({ include: 'test/samples/without_variant_groups.rs' }),
      ],
    })

    const { css } = await uno.generate('')
    expect(css).matchSnapshot()
  })

  it('should generate utils with variant group feature', async () => {
    const uno: UnoGenerator = createGenerator({
      presets: [
        presetWind(),
        presetYew({ include: 'test/samples/with_variant_groups.rs' }),
      ],
    })

    const { css } = await uno.generate('')
    expect(css).matchSnapshot()
  })
})
