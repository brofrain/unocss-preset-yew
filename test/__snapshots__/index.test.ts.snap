// Vitest Snapshot v1

exports[`index > should generate utils with variant group feature 1`] = `
"/* layer: preflights */
*,::before,::after{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgba(0,0,0,0);--un-ring-shadow:0 0 rgba(0,0,0,0);--un-shadow-inset: ;--un-shadow:0 0 rgba(0,0,0,0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgba(147,197,253,0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}::backdrop{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgba(0,0,0,0);--un-ring-shadow:0 0 rgba(0,0,0,0);--un-shadow-inset: ;--un-shadow:0 0 rgba(0,0,0,0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgba(147,197,253,0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}
/* layer: default */
/* layer: default */
.w72{width:18rem;}
.border-1{border-width:1px;border-style:solid;}
.border-blue\\\\/30{border-color:rgba(96,165,250,0.3);}
.rounded{border-radius:0.25rem;}
.bg-green{--un-bg-opacity:1;background-color:rgba(74,222,128,var(--un-bg-opacity));}
.p-x5{padding-left:1.25rem;padding-right:1.25rem;}
.p-y\\\\.5{padding-top:0.125rem;padding-bottom:0.125rem;}
.text-center{text-align:center;}
.placeholder\\\\:text-sm::placeholder{font-size:0.875rem;line-height:1.25rem;}
.fw700{font-weight:700;}
.placeholder\\\\:italic::placeholder{font-style:italic;}
.text-red{--un-text-opacity:1;color:rgba(248,113,113,var(--un-text-opacity));}
.outline-2{outline-width:2px;}
.\\\\!focus\\\\:outline-orange:focus{--un-outline-color-opacity:1 !important;outline-color:rgba(251,146,60,var(--un-outline-color-opacity)) !important;}
.hover\\\\:outline-yellow:hover{--un-outline-color-opacity:1;outline-color:rgba(250,204,21,var(--un-outline-color-opacity));}
.outline-transparent{outline-color:transparent;}
.outline-offset-0{outline-offset:0px;}
.outline{outline-style:solid;}
.transition-all{transition-property:all;transition-timing-function:cubic-bezier(0.4, 0, 0.2, 1);transition-duration:150ms;}"
`;

exports[`index > should generate utils without variant group feature 1`] = `
"/* layer: preflights */
*,::before,::after{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgba(0,0,0,0);--un-ring-shadow:0 0 rgba(0,0,0,0);--un-shadow-inset: ;--un-shadow:0 0 rgba(0,0,0,0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgba(147,197,253,0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}::backdrop{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgba(0,0,0,0);--un-ring-shadow:0 0 rgba(0,0,0,0);--un-shadow-inset: ;--un-shadow:0 0 rgba(0,0,0,0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgba(147,197,253,0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}
/* layer: default */
/* layer: default */
.w72{width:18rem;}
.border-1{border-width:1px;border-style:solid;}
.border-blue\\\\/30{border-color:rgba(96,165,250,0.3);}
.rounded{border-radius:0.25rem;}
.bg-green{--un-bg-opacity:1;background-color:rgba(74,222,128,var(--un-bg-opacity));}
.px5{padding-left:1.25rem;padding-right:1.25rem;}
.py\\\\.5{padding-top:0.125rem;padding-bottom:0.125rem;}
.text-center{text-align:center;}
.placeholder\\\\:text-sm::placeholder{font-size:0.875rem;line-height:1.25rem;}
.fw700{font-weight:700;}
.placeholder\\\\:italic::placeholder{font-style:italic;}
.text-red{--un-text-opacity:1;color:rgba(248,113,113,var(--un-text-opacity));}
.outline-2{outline-width:2px;}
.\\\\!focus\\\\:outline-orange:focus{--un-outline-color-opacity:1 !important;outline-color:rgba(251,146,60,var(--un-outline-color-opacity)) !important;}
.hover\\\\:outline-yellow:hover{--un-outline-color-opacity:1;outline-color:rgba(250,204,21,var(--un-outline-color-opacity));}
.outline-transparent{outline-color:transparent;}
.outline-offset-0{outline-offset:0px;}
.outline{outline-style:solid;}
.transition-all{transition-property:all;transition-timing-function:cubic-bezier(0.4, 0, 0.2, 1);transition-duration:150ms;}"
`;

exports[`index > should retrieve correct classes with variant group feature 1`] = `
Set {
  "w72",
  "p-x5",
  "p-y.5",
  "text-center",
  "text-red",
  "some-fake-class",
  "fw700",
  "bg-green",
  "rounded",
  "border-1",
  "border-blue/30",
  "placeholder:italic",
  "placeholder:text-sm",
  "placeholder:text-secondary/75",
  "outline",
  "outline-2",
  "outline-offset-0",
  "outline-transparent",
  "hover:outline-yellow",
  "!focus:outline-orange",
  "transition-all",
}
`;

exports[`index > should retrieve correct classes without variant group feature 1`] = `
Set {
  "w72",
  "px5",
  "py.5",
  "text-center",
  "some-fake-class",
  "text-red",
  "fw700",
  "bg-green",
  "rounded",
  "border-1",
  "border-blue/30",
  "placeholder:italic",
  "placeholder:text-sm",
  "placeholder:text-secondary/75",
  "outline",
  "outline-2",
  "outline-offset-0",
  "outline-transparent",
  "hover:outline-yellow",
  "!focus:outline-orange",
  "transition-all",
}
`;
